src/Microsoft.ML.Core/Prediction/ITrainer.cs

```diff
@@ -87,91 +81,60 @@ public interface ITrainer
+        /// Convenience train extension for the case where one has only a training set with no auxiliary information.
+        /// Equivalent to calling <see cref="ITrainer.Train(TrainContext)"/>
+        /// on a <see cref="TrainContext"/> constructed with <paramref name="trainData"/>.
         /// </summary>
-        /// <param name="data">Training dataset</param>
-        /// <param name="validData">Validation dataset</param>
-        void Train(TDataSet data, TDataSet validData);
-    }
+        /// <param name="trainer">The trainer</param>
+        /// <param name="trainData">The training data.</param>
+        /// <returns>The trained predictor</returns>
+        public static IPredictor Train(this ITrainer trainer, RoleMappedData trainData)
+            => trainer.Train(new TrainContext(trainData));
 
-    public interface IIncrementalTrainer<in TDataSet, in TPredictor> : ITrainer<TDataSet>
-    {
-        /// <summary>
-        /// Trains a predictor using the specified dataset and a trained predictor.
-        /// </summary>
-        /// <param name="data">Training dataset</param>
-        /// <param name="predictor">A trained predictor</param>
-        void Train(TDataSet data, TPredictor predictor);
-    }
-
-    public interface IIncrementalValidatingTrainer<in TDataSet, in TPredictor> : ITrainer<TDataSet>
-    {
         /// <summary>
-        /// Trains a predictor using the specified dataset and a trained predictor.
+        /// Convenience train extension for the case where one has only a training set with no auxiliary information.
+        /// Equivalent to calling <see cref="ITrainer{TPredictor}.Train(TrainContext)"/>
+        /// on a <see cref="TrainContext"/> constructed with <paramref name="trainData"/>.
         /// </summary>
-        /// <param name="data">Training dataset</param>
-        /// <param name="validData">Validation dataset</param>
-        /// <param name="predictor">A trained predictor</param>
-        void Train(TDataSet data, TDataSet validData, TPredictor predictor);
+        /// <param name="trainer">The trainer</param>
+        /// <param name="trainData">The training data.</param>
+        /// <returns>The trained predictor</returns>
+        public static TPredictor Train<TPredictor>(this ITrainer<TPredictor> trainer, RoleMappedData trainData) where TPredictor : IPredictor
+            => trainer.Train(new TrainContext(trainData));
     }
 
-#if FUTURE
-    public interface IMultiTrainer<in TDataSet, in TFeatures, out TResult> :
-        IMultiTrainer<TDataSet, TDataSet, TFeatures, TResult>
-    {
-    }
-
-    public interface IMultiTrainer<in TDataSet, in TDataBatch, in TFeatures, out TResult> :
-        ITrainer<TDataSet, TFeatures, TResult>
+    /// <summary>
+    /// Interface implemented by the MetalinearLearners base class.
+    /// Used to distinguish the MetaLinear Learners from the other learners
+    /// </summary>
+    public interface IMetaLinearTrainer
```

Comment:
*This* interface needs to go. Its entire reason for existing is *apparently* so that this code here can run this test:

https://github.com/dotnet/machinelearning/blob/650316743d2daa9c45a8a31431b987668b9f0c59/src/Microsoft.ML.PipelineInference/ExperimentsGenerator.cs#L114

has a way to not use a specific type of trainer. But because we dislike that trainer so much we didn't bother to ship it as part of ML.NET, so to keep that check, apparently someone thought it was a brilliant idea to introduce that interface. (I guess?)... then of course put it right into `ITrainer.cs`. So the entire reason why we have that interface and put it in such a central location, I suppose, is because we don't think the trainer it corresponds to is very good. That's *twisted*. :) However, I would prefer to solve it in the following way, by either migrating `MetaLinearPredictor` to ML.NET (and putting the check directly in that pipeline inference code), or annihilating it altogether. But yeah, this marker interface is a bad idea. But I feel like I don't have a good way to do that in *this* PR.

Regarding marker interfaces in general being bad, yeah, they probably ought to be avoided... I'm a little wary though of any sort of blanket policy against them. The common complaint against them is that interfaces are contracts and an interface with no contract is inherently confusing. (And indeed you may have observed in this code-base, we certainly don't shy away from attributes either.) But like anything, you have to weight the confusion caused by that "solution" to a problem, to the alternative.

I am imagining something: let's imagine that we had `IBinaryPredictor` as a marker interface on binary predictors. Then we get to, just in the type system, check whether a trainer is a binary trainer just by doing `trainer is ITrainer<IBinaryPredictor>`. In one fell swoop:

1. we get to get rid of `PredictionKind` (I consider that `enum` a mistake since its existence discourages new types of predictors from being added, e.g., multilabel). We might still have it in some tooling (like a GUI), but as a central concept, it is out.

2. that also means no more `PredictionKind` properties on `IPredictor` and `ITrainer`,

3. we can correspondingly remove `SignatureBinaryClassifierTrainer` since the little tricks 

4. I want to get rid of our internal depenency injection framework, alternate schemes definitely work over types, but they definitely won't work over our custom little attributes marking this or that as binary classification, regression, or whatever,

5. utilities that require specific types of trainers and predictors can ask for them. So for  example, as we've discussed elsewhere, we have the idea of a `OneVersusAll` trainer, that would take a mechanism.

If we could turn this:

https://github.com/dotnet/machinelearning/blob/650316743d2daa9c45a8a31431b987668b9f0c59/src/Microsoft.ML.StandardLearners/Standard/MultiClass/MetaMulticlassTrainer.cs#L26-L27

Into something *like* this:

```csharp
Func<IHostEnvironment, ITrainer<IBinaryPredictor>> TrainerFactor = env => new LinearSvm(env);
```

Then inserting the wrong sort of trainer would immediately result in a compile-time error, not a runtime error.

So are marker interfaces inherently bad, I'm not so convinced of this, since the alternative envisioned is a runtime failure, and one for which we have to write a non-trivial amount of support code to accomodate. But, eh, I might be wrong. Just not sure about that.

Reply:
Just to be clear, I do definitely cringe whenever I see a marker interface (if you like that, you're going to love `IPredictorProducing<...>`). So I think the advice is generally good. I'm just not 100% sold that they aren't sometimes the most elegant, straightforward answer to some problems, and indeed some problems that we actually immediately have.

Reply:
Note: the rule does give an escape hatch for when these are valuable at compile time.

> If the identification must occur at compile time, then it is acceptable to use an empty interface.

Reply:
Turns out we can get rid of it, according to @justinormont thanks Justin.